internal/functions/metricsapi/prepare-new-version.go

```diff
@@ -0,0 +1,252 @@
+}
+
+func getNotificationCounters(ctx context.Context, now time.Time) (*counts, error) {
+	logger := logging.FromContext(ctx)
+
+	var date = now.Add(time.Hour * -24).Format("20060102")
+
+	// yesterday
+
+	yesterdayData, err := getNotificationCounter(ctx, date)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infof("Notifications count for yesterday: %v", yesterdayData.NotificationsCount)
+
+	// total
+
+	totalData, err := getNotificationCounter(ctx, "total")
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infof("Notifications total count: %v", totalData.NotificationsCount)
+
+	return &counts{
+		yesterday: int32(yesterdayData.NotificationsCount),
+		total:     int32(totalData.NotificationsCount),
+	}, nil
+}
+
+func getNotificationCounter(ctx context.Context, key string) (*structs.NotificationCounter, error) {
+	logger := logging.FromContext(ctx)
+	var storeClient = store.Client{}
+
+	logger.Debugf("Getting notification counter with key %v", key)
+
+	doc := storeClient.Doc(constants.CollectionNotificationCounters, key)
+
+	// TODO handle not found
+
+	rec, err := doc.Get(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("Error while querying Firestore: %v", err)
+	}
+
+	var data structs.NotificationCounter
+	err = rec.DataTo(&data)
+	if err != nil {
+		return nil, fmt.Errorf("Error while querying Firestore: %v", err)
+	}
+	return &data, nil
+}
+
+func getActivationCounters(ctx context.Context, now time.Time) (*counts, error) {
+	logger := logging.FromContext(ctx)
+
+	var date = now.Add(time.Hour * -24).Format("20060102")
+
+	// yesterday
+
+	yesterdayData, err := getActivationCounter(ctx, date)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infof("Activations count for yesterday: %v", yesterdayData.UsersCount)
+
+	// total
+
+	totalData, err := getActivationCounter(ctx, "total")
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infof("Activations total count: %v", totalData.UsersCount)
+
+	return &counts{
+		yesterday: int32(yesterdayData.UsersCount),
+		total:     int32(totalData.UsersCount),
+	}, nil
+}
+
+func getActivationCounter(ctx context.Context, key string) (*structs.UserCounter, error) {
+	logger := logging.FromContext(ctx)
+	client := realtimedb.Client{}
+
+	logger.Debugf("Getting activation counter with key %v", key)
+
+	var data structs.UserCounter
+
+	// TODO handle not found
+
+	if err := client.NewRef(constants.DbUserCountersPrefix+key).Get(ctx, &data); err != nil {
+		logger.Debugf("Error while querying DB: %v", err)
+		return nil, err
+	}
+
+	return &data, nil
+}
+
+func getPublishersCount(ctx context.Context) (*counts, error) {
+	logger := logging.FromContext(ctx)
+
+	startOfTomorrow := time.Now().UTC().Add(time.Hour * -24).Truncate(time.Hour * 24)
+	startOfErouska, err := time.Parse("02.01.2006", "01.09.2020")
+	if err != nil {
+		panic(err)
+	}
+
+	projectID, ok := os.LookupEnv("METRICS_PROJECT_ID")
+	if !ok {
+		return nil, fmt.Errorf("Could not find METRICS_PROJECT_ID env")
+	}
```

Comment:
why is this inside of the function and not as an argument?

Reply:
I didn't want to let the value to be passed over too many levels but truth is that it should fail ASAOP if this is not available, so I moved it.

Reply:
that doesn't make any sense to me, this function is doing way too many things, including ENV var handling. Why can't the function be minimal, doing just the one thing, taking in input as an argument - then it is so much easier to mock it, test it, reuse it...

Reply:
You are right but since testing is not possible for many reasons here it's not important to stick with that too tight... :-) But le'ts keep it for the next time :-P